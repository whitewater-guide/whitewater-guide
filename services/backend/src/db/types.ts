export interface RawTimestamped {
  created_at: Date;
  updated_at: Date;
}

export interface WithLanguage {
  language: string;
}

export const timestampResolvers = {
  createdAt: (node: RawTimestamped) => new Date(node.created_at).toISOString(),
  updatedAt: (node: RawTimestamped) => new Date(node.updated_at).toISOString(),
};
