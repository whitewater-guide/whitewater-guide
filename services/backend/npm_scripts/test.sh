#!/usr/bin/env bash

set -o allexport
source ../../config/.env.test
source .env.test
set +o allexport

./node_modules/.bin/jest -i --forceExit "$@"
