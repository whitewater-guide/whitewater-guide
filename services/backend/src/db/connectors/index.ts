export * from './types';
export * from './OffsetConnector';
export * from './createConnectors';
