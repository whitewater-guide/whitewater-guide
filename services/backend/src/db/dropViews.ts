import Knex from 'knex';

export async function dropViews(db: Knex, ...views: string[]) {
  for (const view of views) {
    await db.raw(`DROP VIEW IF EXISTS ${view}_view`);
  }
}
