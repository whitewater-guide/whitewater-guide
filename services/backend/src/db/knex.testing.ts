import knex from 'knex';

const test: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
  },
  pool: {
    min: 0,
    max: 1,
    idleTimeoutMillis: 500,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './dist/migrations',
  } as any,
  seeds: {
    loadExtensions: ['.js'],
    directory: './dist/seeds/test',
  } as any,
  debug: false, // process.env.DATABASE_DEBUG === 'true',
};

module.exports = test;
