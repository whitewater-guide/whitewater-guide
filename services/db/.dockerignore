.env.*
Dockerfile
package.json
/config/x-dev-restore.sh
/config/*.bak
/scripts
/.env
