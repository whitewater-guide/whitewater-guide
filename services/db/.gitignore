.env
!.env.*.secret
/config/*.bak
/config/*.csv
