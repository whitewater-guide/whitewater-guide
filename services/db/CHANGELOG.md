## 0.0.25 (2019-01-31)
