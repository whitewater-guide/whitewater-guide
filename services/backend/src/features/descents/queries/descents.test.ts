import { holdTransaction, rollbackTransaction } from '~/db';
import { fakeContext, runQuery } from '~/test';
import { TEST_USER, TEST_USER2, TEST_USER_ID } from '~/seeds/test/01_users';
import {
  DESCENT_02,
  SECTION_1,
  DESCENT_07,
  DESCENT_03,
  DESCENT_05,
  DESCENT_01,
  DESCENT_10,
  DESCENT_04,
} from '~/seeds/test/18_descents';
import { DescentsFilter } from '@whitewater-guide/commons';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const query = `
  query listDescents($filter: DescentsFilter, $page: Page) {
    descents(filter: $filter, page: $page) {
      edges {
        node {
          id

          startedAt
          duration
          level {
            value
            unit
          }
          comment
          public

          createdAt
          updatedAt

          section {
            id

            region {
              id
              name
            }
            river {
              id
              name
            }
            name
          }
        }
        cursor
      }
      pageInfo {
        endCursor
        hasMore
      }
    }
  }
`;

const getIds = (result: any): string[] =>
  result.descents?.edges.map(({ node }: any) => node.id) || [];

it('should match snapshot', async () => {
  const result = await runQuery(query, {}, fakeContext(TEST_USER));
  expect(result).toMatchSnapshot();
});

it.each([
  ['anon should not see private sections', undefined, false],
  ['non-owner should not see private sections', TEST_USER2, false],
  ['owner should see private sections', TEST_USER, true],
])('%s', async (_, user, visible) => {
  const result = await runQuery(query, {}, user);
  const actual = getIds(result.data).includes(DESCENT_02);
  expect(actual).toBe(visible);
});

type FilterTestCase = [string, DescentsFilter, string[]];
// use snapshot as a baseline
it.each<FilterTestCase>([
  [
    'difficulty',
    { difficulty: [2.5, 3] },
    [DESCENT_07, DESCENT_03, DESCENT_02],
  ],
  [
    'open date range',
    { startDate: new Date('2020-01-03T00:00:00.000Z').toISOString() as any },
    [DESCENT_07, DESCENT_05, DESCENT_03],
  ],
  [
    'closed date range',
    {
      startDate: new Date('2020-01-02T00:00:00.000Z').toISOString() as any,
      endDate: new Date('2020-01-05T00:00:00.000Z').toISOString() as any,
    },
    [DESCENT_05, DESCENT_03, DESCENT_02],
  ],
  [
    'user',
    {
      userId: TEST_USER_ID,
    },
    [DESCENT_03, DESCENT_02, DESCENT_01],
  ],
  [
    'section id', // including derived sections
    {
      sectionId: SECTION_1,
    },
    [DESCENT_10, DESCENT_04, DESCENT_01],
  ],
  [
    'section (full) name',
    {
      sectionName: 'oR',
    },
    [
      DESCENT_07,
      DESCENT_05,
      DESCENT_03,
      DESCENT_02,
      DESCENT_10,
      DESCENT_04,
      DESCENT_01,
    ],
  ],
])(
  `should filter and paginate by %s`,
  async (_, filter, [id1, id2, ...restIds]) => {
    let result = await runQuery(
      query,
      { filter, page: { limit: 1 } },
      TEST_USER,
    );
    expect(result.errors).toBeUndefined();
    expect(getIds(result.data)).toEqual([id1]);
    expect(result.data.descents?.pageInfo.hasMore).toBe(true);
    result = await runQuery(
      query,
      {
        filter,
        page: {
          limit: 1,
          after: result.data.descents?.pageInfo.endCursor,
        },
      },
      TEST_USER,
    );
    expect(result.errors).toBeUndefined();
    expect(getIds(result.data)).toEqual([id2]);
    expect(result.data.descents?.pageInfo.hasMore).toBe(true);
    result = await runQuery(
      query,
      {
        filter,
        page: {
          limit: 99,
          after: result.data.descents?.pageInfo.endCursor,
        },
      },
      TEST_USER,
    );
    expect(result.errors).toBeUndefined();
    expect(getIds(result.data)).toEqual(restIds);
    expect(result.data.descents?.pageInfo.hasMore).toBe(false);
  },
);
