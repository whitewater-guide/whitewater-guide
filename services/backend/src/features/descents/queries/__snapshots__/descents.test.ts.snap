// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "data": Object {
    "descents": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
